src/main/java/com/sourcegraph/cody/agent/CurrentConfigFeatures.java

```diff
@@ -0,0 +1,100 @@
+package com.sourcegraph.cody.agent;
+
+import com.intellij.openapi.components.Service;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * {@link CurrentConfigFeatures} distributes the information about Cody feature configuration, like
+ * whether attribution is turned on or off, so that UI can adapt accordingly.
+ *
+ * <p>These features are turned on/off on the Sourcegraph instance and are fetched periodically by
+ * the agent.
+ *
+ * <p>Observers implementing {@link Consumer<ConfigFeatures>} can {@link #attach} and then will be
+ * notified of each config feature. Note: the observers will be notified irrespective of whether
+ * value of config feature is the same or different from the previous value, and need to
+ * de-duplicate individually if needed.
+ */
+@Service(Service.Level.PROJECT)
+public final class CurrentConfigFeatures implements ConfigFeaturesObserver {
+
+  /** Most recent {@link ConfigFeatures}. */
+  private final AtomicReference<ConfigFeatures> features =
+      new AtomicReference<>(new ConfigFeatures(false));
+
+  /**
+   * Observers that are attached (see {@link #attach}) and receive updates
+   * ({@link ConfigFeaturesObserver#update).
+   *
+   * <p>{@link IdentityObserver} is used here in order to provide precise
+   * dispose semantics (removal from this set) irrespecitve of {@link #equals}
+   * behavior on the delegate {@link ConfigFeaturesObserver}.
+   */
+  private final Set<IdentityObserver> observers = ConcurrentHashMap.newKeySet();
+
+  /** Retrieve the most recent {@link ConfigFeatures} value. */
+  public ConfigFeatures get() {
+    return features.get();
+  }
+
+  /**
+   * New {@link ConfigFeatures} arrive from the agent. This method updates state and notifies all
+   * observers.
+   */
+  @Override
+  public void update(ConfigFeatures configFeatures) {
+    this.features.set(configFeatures);
+    observers.forEach((observer) -> observer.update(configFeatures));
+  }
+
+  /**
+   * Given listener will be given new {@link ConfigFeatures} whenever they arrive. Observation
+   * relationship is ended once the returned cleanup {@link Runnable} is {@link Runnable#run()}.
+   */
+  public Runnable attach(ConfigFeaturesObserver observer) {
```

Comment:
Minor thing, but you could use `CancellationToken` here.

```kotlin
val cancellationToken = CancellationToken()
cancellationToken.onCancel { observers.remove(id) }
return cancellationToken
```

A bit more code but I think it maybe is a bit clear regarding the intent.

Reply:
Done!